cumulusci/core/flowrunner.py

```diff
@@ -0,0 +1,428 @@
+- Task default (i.e. `.tasks.TASKNAME.options`)
+- Flow definition task options (i.e. `.flows.FLOWNAME.steps.STEPNUM.options`)
+- Flow definition subflow options (i.e. `.flows.FLOWNAME.steps.STEPNUM.options.TASKNAME`)
+    see `dev_org_namespaced` for an example
+- Flow runtime (i.e. on the commandline)
+
+
+"""
+
+# we don't actually use this set of imports, they're just in type
+# comments, which require explicit runtime import when checking...
+try:
+    from typing import List
+except ImportError:
+    pass
+
+import copy
+import logging
+from collections import namedtuple
+from distutils.version import LooseVersion
+
+
+from cumulusci.core.config import TaskConfig
+from cumulusci.core.exceptions import FlowConfigError, FlowInfiniteLoopError
+from cumulusci.core.utils import import_class
+
+
+class StepSpec(object):
+    """ simple namespace to describe what the flowrunner should do each step """
+
+    __slots__ = (
+        "step_num",
+        "task_name",
+        "task_options",
+        "allow_failure",
+        "from_flow",
+        "skip",
+    )
+
+    def __init__(
+        self,
+        step_num,
+        task_name,
+        task_options,
+        allow_failure=False,
+        from_flow=None,
+        skip=None,
+    ):
+        self.step_num = step_num
+        self.task_name = task_name
+        self.task_options = task_options
+        self.allow_failure = allow_failure
+        self.from_flow = from_flow
+        self.skip = skip
+
+    def __repr__(self):
+        skipstr = ""
+        if self.skip:
+            skipstr = "!SKIP! "
+        return "<{skip}StepSpec {num}:{name} {cfg}>".format(
+            num=self.step_num, name=self.task_name, cfg=self.task_options, skip=skipstr
+        )
+
+    @property
+    def for_display(self):
+        skip = ""
+        if self.skip:
+            skip = " [SKIP]"
+        return "{step_num:8}: {task_name}{skip}".format(
+            step_num=self.step_num, task_name=self.task_name, skip=skip
+        )
+
+
+StepResult = namedtuple(
+    "StepResult", ["step_num", "task_name", "result", "return_values", "exception"]
+)
+
+
+class TaskRunner(object):
+    def __init__(self, project_config, org_config, runtime=None):
+        self.project_config = project_config
+        self.runtime = runtime
+        self.org_config = org_config
+
+    def run_step(self, step):
+        """
+        Run a step.
+
+        :param step: StepSpec
+        :return: StepResult
+        """
+        # get the base task_config from the project config, as a dict for easier manipulation.
+        task_config = copy.deepcopy(self.project_config.get_task(step.task_name).config)
+        if "options" not in task_config:
+            task_config["options"] = {}
+
+        # update the copied task_config with override options from the flow definition
+        task_config["options"].update(step.task_options)
+
+        # get task implementation class
+        task_class = import_class(task_config["class_path"])
```

Comment:
It would be nice to find out about a misspelled class path at initialization time rather than runtime.